File: Grokking the Coding Interview/Pattern - Sliding Window/fruitsIntoBaskets.js
// longest contiguous subarray with two unique characters

const fruitsIntoBaskets = (arr) => {
  let freq = {};
  let maxLength = -Infinity;
  let windowStart = 0;
  for (let windowEnd = 0; windowEnd < arr.length; windowEnd += 1) {
    if (!freq[arr[windowEnd]]) freq[arr[windowEnd]] = 1;
    else freq[arr[windowEnd]] += 1;
    while (Object.keys(freq).length > 2) {
      freq[arr[windowStart]] -= 1;
      if (freq[arr[windowStart]] === 0) delete freq[arr[windowStart]];
      windowStart += 1;
    }
    maxLength = Math.max(maxLength, windowEnd - windowStart + 1);
  }
  return (maxLength === -Infinity) ? 0 : maxLength;
};

console.log(fruitsIntoBaskets(['A', 'B', 'C', 'A', 'C'])); // -> 3
console.log(fruitsIntoBaskets(['A', 'B', 'C', 'B', 'B', 'C'])); // -> 3

File: Grokking the Coding Interview/Pattern - Sliding Window/fruits_into_baskets.py
def fruits_into_baskets(ls):
    freq = {}
    max_len = float('-inf')
    l = 0
    for r, val in enumerate(ls):
        if val in freq:
            freq[val] += 1
        else: freq[val] = 1
    while len(freq) > 2:
        freq[ls[l]] -= 1
        if freq[ls[l]] == 0:
            del freq[ls[l]]
        l += 1
    max_len = max(max_len, r - l + 1)
    return max_len if max_len != float('-inf') else 0

print(fruits_into_baskets(['A', 'B', 'C', 'A', 'C']))
print(fruits_into_baskets(['A', 'B', 'C', 'B', 'B', 'C']))
